django/db/models/window.py

```diff
@@ -0,0 +1,187 @@
+            window_sql.append(' ORDER BY ')
+            order_sql, order_params = compiler.compile(self.order_by)
+            window_sql.extend(''.join(order_sql))
+            window_params.extend(order_params)
+
+        if self.frame:
+            frame_sql, frame_params = compiler.compile(self.frame)
+            window_sql.extend(' ' + frame_sql)
+            window_params.extend(frame_params)
+
+        params.extend(window_params)
+        template = template or self.template
+
+        return template % {
+            'expression': expr_sql,
+            'window': ''.join(window_sql).strip()
+        }, params
+
+    def __str__(self):
+        return '{} OVER ({}{}{})'.format(
+            str(self.source_expression),
+            'PARTITION BY ' + str(self.partition_by) if self.partition_by else '',
+            'ORDER BY ' + str(self.order_by) if self.order_by else '',
+            str(self.frame or ''),
+        )
+
+    def __repr__(self):
+        return '<%s: %s>' % (self.__class__.__name__, self)
+
+    def get_group_by_cols(self):
+        return []
+
+
+class WindowFrame(Expression):
+    """
+    Model of the frame-clause in use in window expressions. There are two
+    types of frame clauses, which are subclassed from this, however, all the
+    processing and validation (by no means intended complete) is done here.
+
+    Thus, providing an end for a frame is optional (the default is
+    UNBOUNDED FOLLOWING, which is the last row in the frame).
+    """
+    template = '%(frametype)s BETWEEN %(start)s AND %(end)s'
+
+    def __init__(self, start=None, end=None):
+        self.start = start
+        self.end = end
+        super().__init__(output_field=None)
+
+    def copy(self):
+        clone = super().copy()
+        clone.frametype = self.frametype
+        clone.start = self.start
+        clone.end = self.end
+        return clone
+
+    def set_source_expressions(self, exprs):
+        self.start, self.end = exprs
+
+    def get_source_expressions(self):
+        return [self.start, self.end]
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        return self.copy()
+
+    def as_sql(self, compiler, connection):
+        """
+        This should be in operations for each of the database backends that
+        support this. MariaDB 10.2 allows RANGE PRECEDING|FOLLOWING unlike
+        Oracle and PostgreSQL.
+        """
+        connection.ops.check_expression_support(self)
+        start, end = self.window_frame_start_end(connection, self.start, self.end)
+        return self.template % {
+            'frametype': self.frametype,
+            'start': start,
+            'end': end,
+        }, []
+
+    def __repr__(self):
+        return '<%s: %s>' % (self.__class__.__name__, self)
+
+    def __str__(self):
+        if self.start is not None and self.start < 0:
+            start = '%d PRECEDING' % abs(self.start)
```

Comment:
I think you should use constants from `DatabaseOperations` instead of `'PRECEDING'`, `'CURRENT ROW'`, `'UNBOUNDED'` etc.